File: src/server_c/message.c
#include "communication.h"
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include "server.h"

// Function to create a new DepartureTime object
DepartureTime* create_departure_time(int year, int month, int day, int hour, int minute) {
    DepartureTime* departure_time = (DepartureTime*)malloc(sizeof(DepartureTime));
    departure_time->year = year;
    departure_time->month = month;
    departure_time->day = day;
    departure_time->hour = hour;
    departure_time->minute = minute;
    return departure_time;
}

// Function to free a DepartureTime object
void free_departure_time(DepartureTime* departure_time) {
    if (departure_time) {
        free(departure_time);
    }
}

// Function to create a new Flight object
Flight* create_flight(int flight_id, const char* source, const char* destination,
                      DepartureTime* departure_time, float airfare,
                      int seat_availability, int baggage_availability) {
    Flight* flight = (Flight*)malloc(sizeof(Flight));
    flight->flight_id = flight_id;

    // Allocate memory and copy strings for source and destination
    flight->source_place = (char*)malloc(strlen(source) + 1);
    strcpy(flight->source_place, source);

    flight->destination_place = (char*)malloc(strlen(destination) + 1);
    strcpy(flight->destination_place, destination);

    // Copy the pointer for DepartureTime
    flight->departure_time = departure_time;

    flight->airfare = airfare;
    flight->seat_availability = seat_availability;
    flight->baggage_availability = baggage_availability;

    return flight;
}

// Function to free a Flight object, including its strings and DepartureTime
void free_flight(Flight* flight) {
    if (flight) {
        free(flight->source_place);
        free(flight->destination_place);
        free_departure_time(flight->departure_time); // Free DepartureTime
        free(flight);
    }
}

// Function to create a new Message object with Flight data
Message* create_message(uint8_t message_type, uint32_t request_id, Flight* flight_data) {
    Message* message = (Message*)malloc(sizeof(Message));
    message->message_type = message_type;
    message->request_id = request_id;

    // Marshal the Flight data into the data field of the Message
    message->data = marshal_flight(flight_data, &message->data_length);

    return message;
}

// Function to free a Message object
void free_message(Message* message) {
    if (message) {
        free(message->data);
        free(message);
    }
}

File: src/client_java/Message.java
package SC6103_DS.src.client_java;

public class Message {
    private int message_type;
    private int request_id;
    private byte[] payload;

    public Message(int message_type, int request_id, byte[] payload) {
        this.message_type = message_type;
        this.request_id = request_id;
        this.payload = payload;
    }

    public int get_message_type() {
        return message_type;
    }

    public int get_request_id() {
        return request_id;
    }

    public byte[] get_payload() {
        return payload;
    }

    // Nested class for DepartureTime
    public static class DepartureTime {
        private int year;
        private int month;
        private int day;
        private int hour;
        private int minute;

        public DepartureTime(int year, int month, int day, int hour, int minute) {
            this.year = year;
            this.month = month;
            this.day = day;
            this.hour = hour;
            this.minute = minute;
        }

        @Override
        public String toString() {
            return String.format("%d-%02d-%02d %02d:%02d", year, month, day, hour, minute);
        }
    }
}
